### ISSUE TITLE
EndpointRun: Run an endpoint from the CLI

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#624]
Add a new Action that runs an endpoint from the CLI.  Please note this is contingent on issue #576.  Once we figure out #576, this Action will use the appropriate event object and submit it to either a local emulation of your API Gateway endpoint, or your deployed API Gateway endpoint.  This is key to enabling powerful TDD.

**Comment by eahefnawy** [comment]
I'm not sure how a local API emulation would be useful, since we want to test the real AWS APIG. But I love the idea of "invoking" deployed APIG the same way we're invoking deployed lambda functions. This will ensure that our APIG configurations are working as expected.

**Comment by pmuens** [comment]
I think the core advantage is faster speed during TDD development.

I'm also a fan of local testing. It would be great to develop the whole application offline and then push it to AWS.

Not APIG related but recently I stumbled upon the PUT limit of S3 for my free account which I nearly reached because I deployed my function after every tiny change.

**Comment by cangoektas** [comment]
+1 on local testing. Although it is not so often that I don't have access to the internet it happens sometimes. Offline development would be really nice.

**Comment by eahefnawy** [comment]
yep offline development is definitely an advantage, but what we're trying to test here is that **the deployed APIG configurations are working**, if we do local testing, then what are we testing exactly?! if that local test passes, you still have to test it on the real APIG **to make sure your configs were deployed properly**, and that would make local testing useless.

 Local testing makes sense for Functions because you're testing the code quickly, but for APIG, it's the actual deployed APIG configs that you want to test.

What you guys think?

**Comment by remicastaing** [comment]
Personally, I would be happy being able to develop locally and having both a front end and API served with live reload as I know it with node + express.

**Comment by pmuens** [comment]
:+1: Being able to develop locally would be really great.

**Comment by martinmicunda** [comment]
+1

**Comment by cangoektas** [comment]
@eahefnawy I understand that and I agree with you. But for me it just feels very unnatural to not be able to run and test things locally. Personally I am used to run and test first and then deploy. I guess if we don't find a way to make things work locally we just have to adjust to that.

Also about the whole emulating topic. I ask myself, is that really something we could and should do? I can't shake the feeling that we shouldn't be the ones creating local versions of Amazon Web Services. Do you guys know why Amazon hasn't done that already?

**Comment by eahefnawy** [comment]
Thanks for your feedback there guys. It seems that all of you agree that local endpoint testing is a needed feature. So we'll definitely do it!

Could you please provide more feedback on how you think this should work? Because I agree with @cangoektas , I'm not sure if emulating AWS is the best approach. Could you provide some insight on what this feature should cover?

**Comment by vkreal** [comment]
:+1:

**Comment by cangoektas** [comment]
I have an idea but I currently don't have much time to look deeper so I'll just throw this out to you for discussion.

* Enable API Gateway configuration through Swagger files (#283)
* Use Swagger CodeGen to generate server with the defined endpoints
* Configure endpoints to use local functions as handlers

Lots of tools for the above workflow are already there. We could have a look at them and see what we could use:
  - https://github.com/awslabs/aws-apigateway-importer
  - https://github.com/swagger-api/swagger-codegen
  - https://github.com/swagger-api/swagger-js
  - https://github.com/wcandillon/swagger-js-codegen
  - https://github.com/BigstickCarpet/swagger-server

Also, by supporting swagger.json files we would also have a way to document serverless APIs, which would benefit the planned module registry (#524, #525, #575). Although documentation for the registry is another issue we could end up killing two birds with one stone with this approach. What do you think?

**Comment by telson** [comment]
👍

**Comment by woodlandh** [comment]
Was #576 (required for this issue) implemented? I know it was planned but I can't find any follow-up.
Thanks!

**Comment by flomotlik** [comment]
At the moment local running should be supported by the serverless-offline plugin in 0.5

We're currently discussing future development tooling in a separate issue: #1582 

Will close this one in favour of the other ones.
